[SPARK-25636][CORE] spark-submit cuts off the failure reason when there is an error connecting to master

devaraj-kavali opened the pull request:
## What changes were proposed in this pull request?
Cause of the error is wrapped with SparkException, now finding the cause from the wrapped exception and throwing the cause instead of the wrapped exception.

## How was this patch tested?
Verified it manually by checking the cause of the error, it gives the error as shown below.

### Without the PR change

```
[apache-spark]$ ./bin/spark-submit --verbose --master spark://******
....
Error: Exception thrown in awaitResult:
Run with --help for usage help or --verbose for debug output

```

### With the PR change


```
[apache-spark]$ ./bin/spark-submit --verbose --master spark://******
....
Exception in thread "main" org.apache.spark.SparkException: Exception thrown in awaitResult:
        at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:226)
        at org.apache.spark.rpc.RpcTimeout.awaitResult(RpcTimeout.scala:75)
        ....
        at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.io.IOException: Failed to connect to devaraj-pc1/10.3.66.65:7077
        at org.apache.spark.network.client.TransportClientFactory.createClient(TransportClientFactory.java:245)
       ....
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: devaraj-pc1/10.3.66.65:7077
        at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
        ....
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
        ... 1 more
Caused by: java.net.ConnectException: Connection refused
        ... 11 more

```

Commit:
[SPARK-25636][CORE] spark-submit swallows the failure reason when there
is an error connecting to master

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
The only thing I'm confused about is, right now when the `SparkException` is thrown instead of its cause, wouldn't you get a big stack trace starting with the `SparkException`? that `SparkException` is useful too. I don't see why it doesn't print in this case, do you? and can we fix that part.

tgravescs commented:
Please update the title as it seems cut off.

devaraj-kavali commented:
Thanks @srowen  for looking into this.

> ThreadUtils.scala

```
      case NonFatal(t) if !t.isInstanceOf[TimeoutException] =>
        throw new SparkException("Exception thrown in awaitResult: ", t)
```

Here the cause is getting wrapped as SparkException with the message. And in SparkSubmit.scala, it is just printing this message and discarding the caused exception.

> SparkSubmit.scala

```
      override def doSubmit(args: Array[String]): Unit = {
        try {
          super.doSubmit(args)
        } catch {
          case e: SparkUserAppException =>
            exitFn(e.exitCode)
          case e: SparkException =>
            printErrorAndExit(e.getMessage())
        }
```


The other option is to print the whole stack trace instead of just message here. Please let me know your thought, I can make change with this.

srowen commented:
Yeah, that's the issue. Maybe change `printErrorAndExit` to `logError` the exception in this case. It's noisier to be sure but I think losing the stack trace here is significant. @vanzin I think you touched this part last; WDYT?

Otherwise this change doesn't address the fact that you'd still only print a little info about one cause, not all of them.

vanzin commented:
Looking at the 2.3 code, it might be the right thing to just remove the handler for `SparkException` in `SparkSubmit.main`, and let it propagate.

```
    try {
      mainMethod.invoke(null, childArgs.toArray)
    } catch {
      case t: Throwable =>
        findCause(t) match {
          case SparkUserAppException(exitCode) =>
            System.exit(exitCode)

          case t: Throwable =>
            throw t
        }
    }
```

Commit:
Removed handling of SparkException in overridden doSubmit

vanzin commented:
ok to test

SparkQA commented:
**[Test build #97003 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97003/testReport)** for PR 22623 at commit [`81e1730`](https://github.com/apache/spark/commit/81e1730ddf222943eb718e35bb0c6c2599d0335c).

SparkQA commented:
**[Test build #97003 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97003/testReport)** for PR 22623 at commit [`81e1730`](https://github.com/apache/spark/commit/81e1730ddf222943eb718e35bb0c6c2599d0335c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97003/
Test FAILed.

Commit:
Updated the tests according to the change

SparkQA commented:
**[Test build #97021 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97021/testReport)** for PR 22623 at commit [`a82e75f`](https://github.com/apache/spark/commit/a82e75fb4019cf7c0e5ca8279a40e1ac8dbbf53e).

SparkQA commented:
**[Test build #97021 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97021/testReport)** for PR 22623 at commit [`a82e75f`](https://github.com/apache/spark/commit/a82e75fb4019cf7c0e5ca8279a40e1ac8dbbf53e).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97021/
Test FAILed.

SparkQA commented:
**[Test build #4358 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4358/testReport)** for PR 22623 at commit [`a82e75f`](https://github.com/apache/spark/commit/a82e75fb4019cf7c0e5ca8279a40e1ac8dbbf53e).

SparkQA commented:
**[Test build #4358 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4358/testReport)** for PR 22623 at commit [`a82e75f`](https://github.com/apache/spark/commit/a82e75fb4019cf7c0e5ca8279a40e1ac8dbbf53e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

srowen commented on core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:
```diff
@@ -74,20 +74,26 @@ trait TestPrematureExit {
-        case e: Exception => if (!exitedCleanly) throw e
+        case e: Exception =>
+          message = e.getMessage
+          if (!(exitedCleanly || message.contains(searchString))) {
```
If it didn't exit cleanly, can it be possible that the exception is the correct expected one, and that its message contains the search string? I'm probably missing the reason why this has to be checked here.

devaraj-kavali replied:
With this PR change, SparkException will not be caught and thrown directly, and nothing writing to System.err and also no exit(exitedCleanly = false) in this case,  here we need to check the thrown exception message whether it has the expected searchString or not.

vanzin replied:
I actually wonder if throwing this exception below is actually doing anything? `thread.join()` will not throw it in the test thread, which is what would actually matter.

So it seems that to achieve what the comment above says, you should do just this instead:

```
if (!exitedCleanly) {
  message = e.getMessage
}
```

devaraj-kavali replied:
> I actually wonder if throwing this exception below is actually doing anything?

It would give the stack trace in the test log if there is any unexpected exception. With the below, we may give exception message but no stack trace in the log during fail.
```
if (!exitedCleanly) {
  message = e.getMessage
}
```

vanzin replied:
Then just print the exception, or fail the test explicitly with that exception somehow. At least that's explicitly doing what is expected of it.

vanzin replied:
(e.g., by stashing the exception in a variable and doing these checks on the test thread after the `thread.join()` call.)

srowen commented:
That's checked below. Why this change?

devaraj-kavali commented:
This change is to avoid the expected exception is being thrown from the thread and getting printed in the test log.

devaraj-kavali commented:
Yes, you are right, `exitedCleanly` is `false` even when the expected exception is thrown.

Commit:
Refactored the test change

SparkQA commented:
**[Test build #97127 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97127/testReport)** for PR 22623 at commit [`20c4e01`](https://github.com/apache/spark/commit/20c4e01f4fe06f42a84dcc4747fe87e6fd098d1d).

SparkQA commented:
**[Test build #97127 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97127/testReport)** for PR 22623 at commit [`20c4e01`](https://github.com/apache/spark/commit/20c4e01f4fe06f42a84dcc4747fe87e6fd098d1d).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97127/
Test FAILed.

srowen commented on core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:
```diff
@@ -74,20 +74,27 @@ trait TestPrematureExit {
-    if (!joined.contains(searchString)) {
-      fail(s"Search string '$searchString' not found in $joined")
+    val searchStrContainsInEx = exception != null && exception.getMessage.contains(searchString)
+    if(!searchStrContainsInEx){
```
You need spaces around parens here. But I think there's a logic issue; if exception is `null` but `exitedCleanly` is `false`, you try to throw `null`. I get that it won't be null if so, but that highlights that I think they're redundant. And I don't think it's helping to pull out the condition here in a variable.

I think this is just simpler to delete `exitedCleanly` and ...

```
@volatile var: Exception = null
...
if (exception != null) {
  if (!exception.getMessage.contains(searchString)) {
    throw exception
  }
} else if (!joined.contains(searchString) {
  fail(...)
}
```
I think that's equivalent?

devaraj-kavali replied:
I've update as per this, thanks for the suggestion.

Commit:
Refactored the test change

SparkQA commented:
**[Test build #97132 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97132/testReport)** for PR 22623 at commit [`96104bd`](https://github.com/apache/spark/commit/96104bdf0e08cdb498c959dc98d45d3ad896d1b0).

vanzin commented on core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:
```diff
@@ -74,20 +74,28 @@ trait TestPrematureExit {
-    if (!joined.contains(searchString)) {
-      fail(s"Search string '$searchString' not found in $joined")
+    val searchStrContainsInEx = exception != null && exception.getMessage.contains(searchString)
+    if (!searchStrContainsInEx) {
```
I thinks this is a bit hard to follow. Isn't this equivalent to:

```
if (exitedCleanly) {
  // check the output captured in printStream
} else {
  assert(exception != null)
  if (message does not contain search string) {
    throw exception
  }
}
```

Also, now that you moved the logic here, the comment in L83 is kinda orphaned, and a fresher comment explaining this block would be better.

devaraj-kavali replied:
I've updated as per Sean's comment, plz check whether it is ok.

Commit:
Refactored the test change

SparkQA commented:
**[Test build #97133 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97133/testReport)** for PR 22623 at commit [`9e60602`](https://github.com/apache/spark/commit/9e6060289c798117d2e2b18d2f73af1e8eb74d3f).

vanzin commented on core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:
```diff
@@ -70,23 +70,25 @@ trait TestPrematureExit {
+      if (!exception.getMessage.contains(searchString)) {
+        throw exception
+      }
+    } else if (!joined.contains(searchString)) {
```
This could be just `assert(joined.contains(blah))`. You could also move the assignment of `joined` inside the `else` here.

Otherwise looks good pending tests.

SparkQA commented:
**[Test build #97132 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97132/testReport)** for PR 22623 at commit [`96104bd`](https://github.com/apache/spark/commit/96104bdf0e08cdb498c959dc98d45d3ad896d1b0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97132/
Test FAILed.

Commit:
Refactored the test change

SparkQA commented:
**[Test build #97134 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97134/testReport)** for PR 22623 at commit [`71e3e30`](https://github.com/apache/spark/commit/71e3e30aaf94dfb638eca7e2edfc5feddccc0d3d).

SparkQA commented:
**[Test build #97133 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97133/testReport)** for PR 22623 at commit [`9e60602`](https://github.com/apache/spark/commit/9e6060289c798117d2e2b18d2f73af1e8eb74d3f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97133/
Test FAILed.

SparkQA commented:
**[Test build #97134 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97134/testReport)** for PR 22623 at commit [`71e3e30`](https://github.com/apache/spark/commit/71e3e30aaf94dfb638eca7e2edfc5feddccc0d3d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97134/
Test FAILed.

Commit:
MesosClusterDispatcherSuite tests have the case that exitedCleanly=true
and exception is being thrown from the thread, making the change to
support this.

SparkQA commented:
**[Test build #97143 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97143/testReport)** for PR 22623 at commit [`131b3af`](https://github.com/apache/spark/commit/131b3af7ab286d2f001cbf0638cd5dbe7653a420).

SparkQA commented:
**[Test build #97143 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97143/testReport)** for PR 22623 at commit [`131b3af`](https://github.com/apache/spark/commit/131b3af7ab286d2f001cbf0638cd5dbe7653a420).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97143/
Test FAILed.

SparkQA commented:
**[Test build #4362 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4362/testReport)** for PR 22623 at commit [`131b3af`](https://github.com/apache/spark/commit/131b3af7ab286d2f001cbf0638cd5dbe7653a420).

SparkQA commented:
**[Test build #4362 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4362/testReport)** for PR 22623 at commit [`131b3af`](https://github.com/apache/spark/commit/131b3af7ab286d2f001cbf0638cd5dbe7653a420).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #4365 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4365/testReport)** for PR 22623 at commit [`131b3af`](https://github.com/apache/spark/commit/131b3af7ab286d2f001cbf0638cd5dbe7653a420).

vanzin commented on core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:
```diff
@@ -74,20 +74,25 @@ trait TestPrematureExit {
-      fail(s"Search string '$searchString' not found in $joined")
+    if (exitedCleanly) {
+      val joined = printStream.lineBuffer.mkString("\n")
+      assert(joined.contains(searchString), s"Search string '$searchString' not found in $joined")
```
The message is redundant, `assert` already gives you a nice message, e.g.

```
[info]   "foo" did not contain "bar" (blah.scala)
```

vanzin reviewed (commented):
Looks good pending tests.

Commit:
Removed assert redundant message

SparkQA commented:
**[Test build #97179 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97179/testReport)** for PR 22623 at commit [`1002e44`](https://github.com/apache/spark/commit/1002e448dc23bdbb2899893dca840dd4e7f935ec).

SparkQA commented:
**[Test build #97179 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97179/testReport)** for PR 22623 at commit [`1002e44`](https://github.com/apache/spark/commit/1002e448dc23bdbb2899893dca840dd4e7f935ec).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97179/
Test PASSed.
